src/map_field.cpp

```diff
@@ -1719,54 +1672,86 @@ void map::player_in_field( player &u )
-    } else if( player *p = critter.as_player() ) {
-        player_in_field( *p );
+        monster_in_field( *static_cast<monster *>( &critter ), cur_field, inside );
+    } else {
+        if( u && u->in_vehicle ) {
+            // If we are in a vehicle figure out if we are inside (reduces effects usually)
+            // and what part of the vehicle we need to deal with.
+            if( u->in_vehicle ) {
+                if( const optional_vpart_position vp = veh_at( u->pos() ) ) {
+                    if( vp->is_inside() ) {
+                        inside = true;
+                    }
+                }
+            }
+        }
+        player_in_field( *u, cur_field, inside );
     }
 
-    field &curfield = get_field( critter.pos() );
-    for( auto &field_entry_it : curfield ) {
+    for( auto &field_entry_it : cur_field ) {
         field_entry &cur_field_entry = field_entry_it.second;
         if( !cur_field_entry.is_field_alive() ) {
             continue;
         }
         const field_type_id cur_field_id = cur_field_entry.get_field_type();
-        const effect field_fx = cur_field_entry.field_effect();
-        if( field_fx.is_null() ||
-            critter.is_immune_field( cur_field_id ) || field_fx.get_id().is_empty() ||
-            critter.is_immune_effect( field_fx.get_id() ) ) {
-            continue;
-        }
-        player *u = critter.as_player();
-        if( u && cur_field_entry.inside_immune() ) {
-            // If we are in a vehicle figure out if we are inside (reduces effects usually)
-            // and what part of the vehicle we need to deal with.
-            if( u->in_vehicle ) {
-                if( const optional_vpart_position vp = veh_at( u->pos() ) ) {
-                    if( vp->is_inside() ) {
-                        continue;
-                    }
+
+        for( const auto &fe : cur_field_entry.field_effects() ) {
+            if( inside && !one_in( fe.inside_chance ) ) {
+                continue;
+            }
+            if( !inside && !one_in( fe.outside_chance ) ) {
+                continue;
+            }
+            if( inside && fe.inside_immune ) {
+                continue;
+            }
+            if( !inside && fe.outside_immune ) {
+                continue;
+            }
+            const effect field_fx = fe.get_effect();
+            if( field_fx.is_null() ||
+                critter.is_immune_field( cur_field_id ) || field_fx.get_id().is_empty() ||
+                critter.is_immune_effect( field_fx.get_id() ) ) {
+                continue;
+            }
+            bool effect_added;
+            if( fe.is_environmental ) {
+                effect_added = critter.add_env_effect( fe.id, fe.bp, fe.intensity, fe.get_duration() );
+            } else {
+                effect_added = true;
+                critter.add_effect( field_fx );
+            }
+            if( effect_added ) {
+                if( u ) {
```

Comment:
This branching is not needed. The call to `add_msg_player_or_npc` handles the case of `critter==g->u` just fine.